src/Base/Console.h

```diff
@@ -667,11 +746,150 @@ class BaseExport LogLevel {
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::Warning( const char * pMsg, Args&&... args )
+{
+    Warning(std::string(""), pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::Warning( const std::string & notifier, const char * pMsg, Args&&... args )
+{
+    Send<Base::LogStyle::Warning>(notifier, pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::Error( const char * pMsg, Args&&... args )
+{
+    Error(std::string(""), pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::Error( const std::string & notifier, const char * pMsg, Args&&... args )
+{
+    Send<Base::LogStyle::Error>(notifier, pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::CriticalMessage( const char * pMsg, Args&&... args )
+{
+    CriticalMessage(std::string(""), pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::CriticalMessage( const std::string & notifier, const char * pMsg, Args&&... args )
+{
+    Send<Base::LogStyle::CriticalMessage>(notifier, pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::UserNotification( const char * pMsg, Args&&... args )
+{
+    UserNotification(std::string(""), pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::UserNotification( const std::string & notifier, const char * pMsg, Args&&... args )
+{
+    Send<Base::LogStyle::Notification>(notifier, pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::UserTranslatedNotification( const char * pMsg, Args&&... args )
+{
+    UserTranslatedNotification(std::string(""), pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::UserTranslatedNotification( const std::string & notifier, const char * pMsg, Args&&... args )
+{
+    Send<Base::LogStyle::TranslatedNotification>(notifier, pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::Log( const char * pMsg, Args&&... args )
+{
+    Log(std::string(""), pMsg, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void Base::ConsoleSingleton::Log( const std::string & notifier, const char * pMsg, Args&&... args )
+{
+    Send<Base::LogStyle::Log>(notifier, pMsg, std::forward<Args>(args)...);
+}
+
+template <Base::LogStyle category, typename... Args>
+void Base::ConsoleSingleton::Send( const std::string & notifiername, const char * pMsg, Args&&... args )
+{
+    std::string format = fmt::sprintf(pMsg, args...);
+
+    if (connectionMode == Direct) {
+        Notify<category>(notifiername,format);
+    }
+    else {
+        FreeCAD_ConsoleMsgType type;
+
+        if constexpr(category == Base::LogStyle::Log) {

```

Comment:
An alternative of using the if-else blocks is to use an unordered_map that you have to fill once where LogStyle is the key. But this cannot be declared as constexpr. But then std::array can be used which supports constexpr since C++17. IN order to access the element the category must be cast to an init or size_t.

Reply:
This is how it could be written with std::array. Important is to add the types in right order
```
constexpr std::array msgTypes {
    FreeCAD_ConsoleMsgType::MsgType_Wrn,
    FreeCAD_ConsoleMsgType::MsgType_Txt,
    FreeCAD_ConsoleMsgType::MsgType_Err,
    FreeCAD_ConsoleMsgType::MsgType_Log,
    FreeCAD_ConsoleMsgType::MsgType_CriticalTxt,
    FreeCAD_ConsoleMsgType::MsgType_Notification,
    FreeCAD_ConsoleMsgType::MsgType_TranslatedNotification
};

FreeCAD_ConsoleMsgType type = msgTypes.at(static_cast<std::size_t>(category));
```

Reply:
This I will encapsulate in a separate function:

```
constexpr ConsoleSingleton::FreeCAD_ConsoleMsgType ConsoleSingleton::getConsoleMsg(Base::LogStyle style)
{
    constexpr std::array msgTypes { // In order of Base::LogStyle
        FreeCAD_ConsoleMsgType::MsgType_Wrn,
        FreeCAD_ConsoleMsgType::MsgType_Txt,
        FreeCAD_ConsoleMsgType::MsgType_Err,
        FreeCAD_ConsoleMsgType::MsgType_Log,
        FreeCAD_ConsoleMsgType::MsgType_CriticalTxt,
        FreeCAD_ConsoleMsgType::MsgType_Notification,
        FreeCAD_ConsoleMsgType::MsgType_TranslatedNotification
    };

    return msgTypes.at(static_cast<std::size_t>(style));
}
```